### PR TITLE
Generalizability Code (Cucumber Method Patient ID Sampling and Single Trial)

### PR DESCRIPTION
**Pull request by hmathlee**
The following changes to the lung-sliding-classifier repository are proposed:

- generalizability.py. The file will create a new folder called generalizability, reserved for fine-tuning experiments with multi-center clip data. This folder will hold a similar structure to results/data, except each subfolder contains its own subfolder corresponding to one of the multi-center locations (e.g. Chile, Ottawa, Edmonton).

- show_m_modes.py. Save the M-mode images by running this file.

- In config.yml, a new section for generalizability will be created. The uppermost key is called GENERALIZE. Contains parameters for fold sampling and fine-tuning, as well as model metric thresholds for fine-tuning.

### CODE REVIEW

File: src/generalizability.py [hunk 1]
```diff
@@ -0,0 +1,472 @@
...
+            if df['precision'].values[0] < thresholds['PRECISION']:
+                return False
+            if df['true_negatives'].values[0] < thresholds['TRUE_NEGATIVES']:
+                return False
+            if df['true_positives'].values[0] < thresholds['TRUE_POSITIVES']:
+                return False
+            if df['false_negatives'].values[0] > thresholds['FALSE_NEGATIVES']:
+                return False
+            if df['true_positives'].values[0] > thresholds['FALSE_POSITIVES']:
+                return False
```
Comment (jhargun) [inline src/generalizability.py:183]: This feels overly long. Maybe replace with something like this:

```
# Return false if values less than these thresholds
lt_thresh = ['accuracy', 'auc', 'recall', 'precision', 'true_negatives', 'true_positives']
# Return false if values greater than these thresholds
gt_thresh = ['false_negatives', 'true_positives']

for thresh in lt_thresh:
    if df[thresh].values[0] < thresholds[thresh.upper()]:
        return False
for thresh in gt_thresh:
    if df[thresh].values[0] > thresholds[thresh.upper()]:
        return False

return True
```

This should do the same thing but is a bit more readable and has the added advantage of making it easier to add new parameters in the future. Also, is there any reason the function definition is inside the definition of finetune_single_trial? It seems like it would be better as a static function of the class so it can be used elsewhere as well.

File: src/generalizability.py [hunk 2]
```diff
@@ -0,0 +1,472 @@
...
+            fold_list.append(cur_fold)
+            cur_fold = []
+
+        print("Retrieved {} external patient folds. The lengths of each fold are: ".format(len(fold_list)),
+              [len(fold) for fold in fold_list])
+        print('Reading external clip filenames and labels into dataframe...')
+
+        # set up the labelled external dataframe for augmenting the training data
+        ext_dfs = []
+        for center in ['chile', 'ottawa']:
```
Comment (jhargun) [inline src/generalizability.py:235]: Shouldn't this be read from cfg?

File: src/generalizability.py [hunk 3]
```diff
@@ -0,0 +1,472 @@
...
+                tensorboard_path = cfg_full['TRAIN']['PATHS']['TENSORBOARD']
+                if not os.path.exists(tensorboard_path):
+                    os.makedirs(tensorboard_path)
+
+                # Log metrics
+                log_dir = label_file(cfg_full['TRAIN']['PATHS']['TENSORBOARD'])
+
+                # uncomment line below to include tensorboard profiler in callbacks
+                # basic_call = tf.keras.callbacks.TensorBoard(log_dir=log_dir, histogram_freq=1, profile_batch=1)
+                basic_call = tf.keras.callbacks.TensorBoard(log_dir=log_dir, histogram_freq=1)
```
Comment (jhargun) [inline src/generalizability.py:373]: Might be worth adding the profile_batch argument. It doesn't take too much storage or hurt performance but can be good if you want to diagnose where slowdowns are occurring or whether memory is an issue in the future.

**Review by jhargun — COMMENTED** [review]
Looks good to me! Just left a few comments about minor fixes that you may want to implement

File: src/generalizability.py [hunk 4]
```diff
@@ -0,0 +1,472 @@
+import os
+from src.preprocessor import MModePreprocessor
+from models.models import get_model
+from train import *
+from predict import *
+from src.custom.metrics import *
```
Comment (farazali7) [inline src/generalizability.py:6]: NP: It's good practice to optimize imports instead of using `*`

File: src/generalizability.py [hunk 5]
```diff
@@ -0,0 +1,472 @@
...
+from src.data.utils import refresh_folder
+from tensorflow.keras.metrics import Precision, Recall, AUC, TrueNegatives, TruePositives, FalseNegatives, \
+    FalsePositives, Accuracy, SensitivityAtSpecificity
+from tensorflow_addons.metrics.f_scores import F1Score
+
+import tensorflow as tf
+
+cfg = yaml.full_load(open(os.path.join(os.getcwd(), "config.yml"), 'r'))['GENERALIZE']
+cfg_full = yaml.full_load(open(os.path.join(os.getcwd(), "config.yml"), 'r'))
```
Comment (farazali7) [inline src/generalizability.py:15]: This gets a bit confusing later on in parts of the code, it'd be better to keep just the full cfg since an extra `['GENERALIZE']` is the only difference in typing.

File: src/generalizability.py [hunk 6]
```diff
@@ -0,0 +1,472 @@
...
+        # of patient ids, this eliminates the need to dynamically compute each fold size.
+        n_folds = cfg['FOLD_SAMPLE']['NUM_FOLDS']
+        fold_sizes = [len(sliding_ids) // n_folds] * n_folds
+        fold_sizes[-1] += len(sliding_ids) % n_folds
+
+        # Note: the code block above assumes that, if the number of sliding ids is not perfectly divisible by the
+        # desired number of folds, then we make one fold slightly bigger than the rest. This must be modified if
+        # we choose to keep the smaller "leftover" fold.
+
+        # f = open(folds_path, "a")
```
Comment (farazali7) [inline src/generalizability.py:106]: Are these comments safe to be removed?

File: src/generalizability.py [hunk 7]
```diff
@@ -0,0 +1,472 @@
...
+            for i in range(size):
+                cur_id = sliding_ids[-1]
+                sliding_ids = sliding_ids[:-1]
+                cur_fold.append(cur_id)
+                clips_by_id = sliding_df[sliding_df['patient_id'] == cur_id]
+                sliding_count += len(clips_by_id)
+                # f.write(cur_id + '\n')
+            folds.append(cur_fold)
+            cur_fold = []
+            cur_fold_num += 1
```
Comment (farazali7) [inline src/generalizability.py:125]: Looks like this var isn't used or printed anywhere

File: src/generalizability.py (hunk 1 above)
Comment (farazali7) [reply src/generalizability.py:183]: Agree with moving it out of the outer function. We can also rename it to something like `is_performance_ok` or `is_performance_met` to follow standard practice for functions that return bool.

File: src/generalizability.py [hunk 8]
```diff
@@ -0,0 +1,472 @@
...
+        ext_df = pd.concat(ext_dfs, ignore_index=True)
+        labelled_ext_df_path = os.path.join(os.getcwd(), 'generalizability/csvs/labelled_external_dfs/')
+        if cfg['REFRESH_FOLDERS']:
+            refresh_folder(labelled_ext_df_path)
+        labelled_ext_df_path += label_file('labelled_ext')
+        ext_df.to_csv(labelled_ext_df_path)
+
+        train_df = pd.DataFrame([])
+
+        print('Setup complete; cucumber slices ready! Don\'t forget hummus :)')
```
Comment (farazali7) [inline src/generalizability.py:248]: Nice

**Review by farazali7 — COMMENTED** [review]
A lot of great work, Hoseok! Left some minor comments.

File: src/generalizability.py (hunk 1 above)
Comment (hmathlee) [reply src/generalizability.py:183]: Good idea. I agree that having it in the outer function can increase usability elsewhere. Will make that change when writing the outer function code for sure.

File: src/generalizability.py (hunk 2 above)
Comment (hmathlee) [reply src/generalizability.py:235]: Yep! Thanks for the catch.

File: src/generalizability.py (hunk 7 above)
Comment (hmathlee) [reply src/generalizability.py:125]: Yeah, for some reason I had opened the path to the .txt file where I save the patient data slices twice. As per your previous comment, I'm pretty sure those comments are safe to delete. Good catch

File: src/generalizability.py [hunk 9]
```diff
@@ -0,0 +1,472 @@
...
+            print('Fine-tune cross-validation commencing...')
+            for cv_run in range(k):
+                # pick up training from where it left off at the end of previous cross val run.
+                if cv_run > 0:
+                    prev_model_path = os.path.join(cfg['PATHS']['MODEL_OUT'], os.listdir(cfg['PATHS']['MODEL_OUT'])[-1])
+                # for run 0 of cross val, retrieve the model from its location immediately after initial training.
+                else:
+                    prev_model_path = os.path.join(cfg_full['TRAIN']['PATHS']['MODEL_OUT'],
+                                                   os.listdir(cfg_full['TRAIN']['PATHS']['MODEL_OUT'])[-1])
+                model = keras.models.load_model(prev_model_path, compile=False)
```
Comment (bvanberl) [inline src/generalizability.py:330]: Just confirming: the model loaded here is the very same model from the beginning of the cucumber trials? The comments above seem to indicate otherwise.

File: src/generalizability.py [hunk 10]
```diff
@@ -0,0 +1,472 @@
...
+
+                print('FINE-TUNING MODEL FOR FOLD ' + str(cv_run) + ' OF EXTERNAL CLIP TRAINING SET')
+
+                # Set up the train and holdout for each cross-validation run. Note that both parts are taken from the
+                # external data set aside for training. We have a separate dataframe to store testing data.
+                df_val_part = train_folds[cv_run]
+                df_train_part = train_df[~train_df.isin(df_val_part['patient_id'].unique())]
```
Comment (bvanberl) [inline src/generalizability.py:337]: Can you confirm that this doesn't also select the folds *after* the validation fold too?

File: src/generalizability.py [hunk 11]
```diff
@@ -0,0 +1,472 @@
...
+                log_dir = label_file(cfg_full['TRAIN']['PATHS']['TENSORBOARD'])
+
+                # uncomment line below to include tensorboard profiler in callbacks
+                # basic_call = tf.keras.callbacks.TensorBoard(log_dir=log_dir, histogram_freq=1, profile_batch=1)
+                basic_call = tf.keras.callbacks.TensorBoard(log_dir=log_dir, histogram_freq=1)
+
+                # Learning rate scheduler & logging LR
+                writer1 = tf.summary.create_file_writer(log_dir + '/train')
+
+                def scheduler(epoch, lr):
```
Comment (bvanberl) [inline src/generalizability.py:378]: Would it be possible to refactor this helper method and the one in train.py?

File: src/generalizability.py [hunk 12]
```diff
@@ -0,0 +1,472 @@
...
+                        metric_values = history.history[metric]
+                        metrics_df.loc[cv_run, metric] = sum(metric_values) / len(metric_values)
+                gc.collect()
+                tf.keras.backend.clear_session()
+                del model
+
+            # # Record mean and standard deviation of test set results
+            # for metric in metrics:
+            #     metrics_df[metric][n_folds] = metrics_df[metric][0:-2].mean()
+            #     metrics_df[metric][n_folds + 1] = metrics_df[metric][0:-2].std()
```
Comment (bvanberl) [inline src/generalizability.py:443]: Can we delete this?

**Review by bvanberl — COMMENTED** [review]
Fantastic work Hoseok!

File: src/generalizability.py (hunk 11 above)
Comment (hmathlee) [reply src/generalizability.py:378]: The scheduler function must take as parameters an epoch and learning rate. However, in the main body for the single trial code, hparams are supplied, and a TensorFlow writer for logging are defined, both of which are unavailable outside the scope of the CucumberTrainer class. Here is my current workaround (make a "parent" function that returns a scheduler function based on hparams and writer):

File: src/generalizability.py [hunk 13]
```diff
@@ -0,0 +1,472 @@
...
+            print(len(add_set))
+            add_df = ext_df[ext_df['patient_id'].isin(add_set)]
+
+            # Remove the augmentation data from the external dataset. Becomes the new testing set.
+            ext_df = ext_df[~ext_df['patient_id'].isin(add_set)]
+            print('\n{} records removed from the external dataset for training.'.format(len(add_df)))
+
+            # Augment the training data.
+            train_df = pd.concat([train_df, add_df])
+            train_df.sample(frac=1)
```
[the remark this answers is not in the document]
Comment (hmathlee) [reply src/generalizability.py:307]: Yes. I will add a comment :)

File: src/generalizability.py [hunk 14]
```diff
@@ -0,0 +1,472 @@
...
+            model = keras.models.load_model(cfg_full['PREDICT']['MODEL'], compile=False)
+            pred_labels, probs = predict_set(model, ext_df)
+            # Can change these metrics as necessary
+            metrics = [Accuracy(name='accuracy'), AUC(), F1Score(num_classes=2, average='micro', threshold=0.5), Precision(name='precision'), Recall(name='recall'),
+                       TrueNegatives(), TruePositives(), FalseNegatives(), FalsePositives(), Specificity(name='specificity'),
+                       PhiCoefficient(), SensitivityAtSpecificity(0.95)]
+
+            # Create a DataFrame containing metric results from predictions.
+            results = []
+            metric_names = []
```
[the remark this answers is not in the document]
Comment (hmathlee) [reply src/generalizability.py:268]: I'll look into it. For now, I added this comment: "Probabilities are stored as floats, while Accuracy metric compares integer values for labels and probs. Round probabilities to the closer value of 0 or 1." Let me know if that should suffice.

File: src/generalizability.py (hunk 5 above)
Comment (hmathlee) [reply src/generalizability.py:15]: Update: I have left a comment at the top of generalizability.py, explaining the use of cfg versus cfg_full.

**Review by bvanber — APPROVED** [review]

File: src/generalizability.py [hunk 15]
```diff
@@ -0,0 +1,472 @@
...
+        print("{} unique no-sliding patient ids found.\n".format(len(no_sliding_ids)))
+
+        # Shuffle the patient ids.
+        np.random.seed(cfg_full['GENERALIZE']['FOLD_SAMPLE']['SEED'])
+        np.random.shuffle(sliding_ids)
+        np.random.shuffle(no_sliding_ids)
+
+        # Specify the path to this file based on the config file.
+        folds_dir = cfg['PATHS']['FOLDS']
+        if not os.path.exists(folds_dir):
```
Comment (bl21812) [inline src/generalizability.py:76]: option to refresh folder?

File: src/generalizability.py [hunk 16]
```diff
@@ -0,0 +1,472 @@
...
+    external_data_dir = os.path.join(cfg['PATHS']['CSV_OUT'], 'combined_external_data')
+    external_df_path = os.path.join(external_data_dir, label_file('all_external_clips') + '.csv')
+    if cfg['REFRESH_FOLDERS']:
+        refresh_folder(external_data_dir)
+    external_df.to_csv(external_df_path)
+    print("All external clips consolidated into", external_df_path, '\n')
+    # Construct a CucumberTrainer instance with the dataframe.
+    ct = CucumberTrainer(external_df, 5)
+    ct.sample_folds()
+    ct.finetune_single_trial()
```
Comment (bl21812) [inline src/generalizability.py:472]: newline pls :)

File: src/generalizability.py [hunk 17]
```diff
@@ -0,0 +1,472 @@
...
+
+            cur_fold_num += 1
+            num_folds_added += 1
+
+
+if __name__=='__main__':
+    # Consolidate all external clip data into a single dataframe.
+    csvs_dir = cfg['PATHS']['CSV_OUT']
+    external_data = []
+    for center in ['chile', 'ottawa']:
```
Comment (bl21812) [inline src/generalizability.py:457]: every instance of ['chile', 'ottawa'] should be reading from cfg right?

File: config.yml [hunk 18]
```diff
@@ -27,14 +27,14 @@ PREPROCESS:
   # The paths to save files. After running the preprocess pipeline, the npzs and csvs are usually all that are needed to train.
   # You shouldn't *have* to edit these, as they are stored as relative and not absolute paths.
   PATHS:
-    UNMASKED_VIDEOS: 'raw_videos/'  # Path to the videos that are downloaded from the dataset and stored without transformation.
-    MASKED_VIDEOS: 'masked_videos/'  # Path to the videos that result from running the masking tool on UNMASKED_VIDEOS.
-    SMOOTHED_VIDEOS: 'smoothed_videos/'  # Path to the videos that result from applying a median filter to MASKED_VIDEOS (optional).
+    UNMASKED_VIDEOS: 'C:\Users\hsmle\Documents\Deep_Breathe\results/data\raw_videos/'  # Path to the videos that are downloaded from the dataset and stored without transformation.
+    MASKED_VIDEOS: 'C:\Users\hsmle\Documents\Deep_Breathe\lung-sliding-classifier\results/data\masked_videos/'  # Path to the videos that result from running the masking tool on UNMASKED_VIDEOS.
```
[the remark this answers is not in the document]
Comment (bvanberl) [reply config.yml:31]: For our develop branch, this might not be a bad idea. It just makes things look a little cleaner. Future devs can still change to absolute paths (I personally usually put my data on a whole separate drive!)

File: src/data/show_m_modes.py [hunk 19]
```diff
@@ -0,0 +1,56 @@
+from PIL import Image
+import os
+import numpy as np
+import yaml
+import subprocess
+from moviepy.editor import *
+import cv2
+
+npz_dir = 'C:\\Users\\hsmle\\Documents\\Deep_Breathe\\lung-sliding-classifier\\results\\data\\npzs\\'
+mmode_dir = 'C:\\Users\\hsmle\\Documents\\Deep_Breathe\\lung-sliding-classifier\\results\\data\\mmodes'
```
[the remark this answers is not in the document]
Comment (hmathlee) [reply src/data/show_m_modes.py:10]: These were put in the config file! :)

File: src/generalizability.py (hunk 17 above)
Comment (hmathlee) [reply src/generalizability.py:457]: Yes, thanks for catching that

File: src/generalizability.py [hunk 20]
```diff
@@ -0,0 +1,472 @@
...
+    Labels a file name with useful information about date and time.
+    :param file: name of file as string
+    :return labelled_filename: name of file joined with date and time info.
+    '''
+    cur_date = datetime.datetime.now().strftime('%Y-%m-%d_%H-%M-%S')
+    labelled_filename = file + '_' + cur_date
+    return labelled_filename
+
+
+class CucumberTrainer:
```
[the remark this answers is not in the document]
Comment (hmathlee) [reply src/generalizability.py]: Looks like we have a much better name now :) As a reminder we are going with "Threshold-Aware Accumulative Fine-Tuner" (TAAFT).

File: src/generalizability.py (hunk 9 above)
Comment (hmathlee) [reply src/generalizability.py]: The model is retrieved from the previous model path between fine-tuning runs. "Fine-tuning run" would refer to a cross-val run, since fine-tuning is done as cross-validation on the current training set.

File: src/data/show_m_modes.py [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+cfg = yaml.full_load(open(os.path.join(os.getcwd(), "..\\..\\config.yml"), 'r'))['PREPROCESS']
+
+if not os.path.exists(mmode_dir):
+    os.makedirs(mmode_dir)
```
[the remark this answers is not in the document]
Comment (hmathlee) [reply src/data/show_m_modes.py]: Thanks; just added!
